### PR TITLE
Static type inspection.

### PR DESCRIPTION
**Pull request by JordanBoltonMN**
Extends the existing POC for static type inspection.

### CODE REVIEW

File: src/inspection/type/tasks.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+import { Type } from "../../type";
+import { ScopeById, ScopeItemByKey } from "../scope";
+import { getOrCreateScope, getOrCreateType, inspectXorNode } from "./inspectType";
+import { ScopeTypeById, ScopeTypeByKey, TypeInspectionState } from "./type";
+
+export type TriedScopeType = Result<ScopeTypeByKey, CommonError.CommonError>;
+
+export type TriedType = Result<Type.TType, CommonError.CommonError>;
+
+export function tryScopeType(
```
Comment (JordanBoltonMN) [inline src/inspection/type/tasks.ts:17]: Returns typing for each item in the given scope.

File: src/inspection/type/tasks.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        deltaTypeById: new Map(),
+        nodeIdMapCollection,
+        leafNodeIds,
+        scopeById: maybeScopeById !== undefined ? maybeScopeById : new Map(),
+    };
+
+    return ResultUtils.ensureResult(getLocalizationTemplates(settings.locale), () => inspectScopeType(state, nodeId));
+}
+
+export function tryType(
```
Comment (JordanBoltonMN) [inline src/inspection/type/tasks.ts:42]: Returns typing for a single node.

File: src/parser/nodeIdMap/nodeIdMapIterator.ts [hunk 3]
```diff
@@ -141,6 +112,103 @@ export function letKeyValuePairs(
     return maybeArrayWrapper === undefined ? [] : keyValuePairs(nodeIdMapCollection, maybeArrayWrapper);
 }
 
+export function fieldProjectionFieldSelectors(
```
Comment (JordanBoltonMN) [inline src/parser/nodeIdMap/nodeIdMapIterator.ts:115]: Several new helper functions for iterating over TXorNodes.

File: src/parser/parsers/naive.ts [hunk 4]
```diff
@@ -735,7 +735,7 @@ export function readRecursivePrimaryExpression<S extends IParserState = IParserS
     const recursiveArrayNodeKind: Ast.NodeKind.ArrayWrapper = Ast.NodeKind.ArrayWrapper;
     IParserStateUtils.startContext(state, recursiveArrayNodeKind);
 
-    const recursiveExpressions: Ast.TRecursivePrimaryExpression[] = [];
+    const recursiveExpressions: (Ast.InvokeExpression | Ast.ItemAccessExpression | Ast.TFieldAccessExpression)[] = [];
```
Comment (JordanBoltonMN) [inline src/parser/parsers/naive.ts:738]: Stricter typing.

File: src/parser/parsers/naive.ts [hunk 5]
```diff
@@ -1368,7 +1370,7 @@ export function readFieldSpecificationList<S extends IParserState = IParserState
     );
     const fields: Ast.ICsv<Ast.FieldSpecification>[] = [];
     let continueReadingValues: boolean = true;
-    let maybeOpenRecordMarkerConstant: Ast.IConstant<Ast.MiscConstantKind.Ellipsis> | undefined = undefined;
+    let isOnOpenRecordMarker: boolean = false;
```
Comment (JordanBoltonMN) [inline src/parser/parsers/naive.ts:1373]: Bug fix where the open record constant marked as the child of the inccorrect node.

File: src/type/inspector/functionExpressionInspector.ts [hunk 6]
```diff
@@ -93,95 +107,3 @@ function functionParameterXorNodes(
         ? []
         : NodeIdMapIterator.arrayWrapperCsvXorNodes(nodeIdMapCollection, maybeWrappedContent);
 }
-
-function examineParameter(
```
Comment (JordanBoltonMN) [inline src/type/inspector/functionExpressionInspector.ts:97]: moved to TypeUtils

File: src/type/type.ts [hunk 7]
```diff
@@ -1,12 +1,47 @@
...
 
-export type TType = PrimitiveType | TExtendedType;
-export type TExtendedType = AnyUnion | DefinedRecord | DefinedTable;
-export type TExtendedTypeKind = TypeKind.Any | TypeKind.Record | TypeKind.Table;
+export type TType = IPrimitiveType | TExtendedType;
+export type TExtendedType =
+    | AnyUnion
+    | DefinedFunction
+    | DefinedList
+    | DefinedListType
```
Comment (mattmasson) [inline src/type/type.ts:9]: what's the difference between `DefinedList` and `DefinedListType`?

**Pull request merged**